fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/RecordCursorTest.java

```diff
@@ -1012,21 +1014,97 @@ public void testFlatMapReasons() throws Exception {
+        for (int i = 0; i < 3; i++) { // three rounds with no results
+            nextResult = cursor.getNext();
+            assertEquals(RecordCursor.NoNextReason.TIME_LIMIT_REACHED, nextResult.getNoNextReason());
+            cursor = getFilteredFakeOutOfBandCursor(longList, 3, 10, nextResult.getContinuation().toBytes());
+        }
+
+        // Choose the inner branch
+        assertEquals(ImmutableList.of(11, 12), cursor.asList().join());
+        nextResult = cursor.getNext();
+        assertEquals(RecordCursor.NoNextReason.TIME_LIMIT_REACHED, nextResult.getNoNextReason());
+
+        byte[] continuation = nextResult.getContinuation().toBytes();
+        while (continuation != null) {
+            cursor = getFilteredFakeOutOfBandCursor(longList, 3, 10, continuation);
+            nextResult = cursor.getNext();
+            continuation = nextResult.getContinuation().toBytes();
+        }
+    }
+
+    @Test
+    public void orElseContinueWithElseBranchAfterDecision() {
+        final List<Integer> innerList = Arrays.asList(1, 2, 3, 4, 5);
+        final List<Integer> elseList = Arrays.asList(-1, -2, -3, -4, -5);
+        final BiFunction<Executor, byte[], RecordCursor<Integer>> orElse = (x, cont) ->
+                new FakeOutOfBandCursor<>(RecordCursor.fromList(elseList, cont), 3, RecordCursor.NoNextReason.TIME_LIMIT_REACHED);
+
+        // Go through inner cursor to determine that it's empty.
+        RecordCursor<Integer> cursor = RecordCursor.orElse(cont -> new FakeOutOfBandCursor<>(RecordCursor.fromList(innerList, cont), 3)
+                .filter(i -> false), orElse, null);
+        assertEquals(Collections.emptyList(), cursor.asList().join());
+        RecordCursorResult<Integer> nextResult = cursor.getNext();
+        assertEquals(RecordCursor.NoNextReason.TIME_LIMIT_REACHED, nextResult.getNoNextReason());
+        cursor = RecordCursor.orElse(cont -> new FakeOutOfBandCursor<>(RecordCursor.fromList(innerList, cont), 3)
+                .filter(i -> false), orElse, nextResult.getContinuation().toBytes());
+
+        // Switch to else branch, but get stuck with an out-of-band limit.
+        assertEquals(ImmutableList.of(-1, -2, -3), cursor.asList().join());
+        nextResult = cursor.getNext();
+        assertEquals(RecordCursor.NoNextReason.TIME_LIMIT_REACHED, nextResult.getNoNextReason());
+
+        // Continue else branch
+        cursor = RecordCursor.orElse(cont -> new FakeOutOfBandCursor<>(RecordCursor.fromList(innerList, cont), 3)
+                .filter(i -> false), orElse, nextResult.getContinuation().toBytes());
+        assertEquals(ImmutableList.of(-4, -5), cursor.asList().join());
+        nextResult = cursor.getNext();
+        assertEquals(RecordCursor.NoNextReason.SOURCE_EXHAUSTED, nextResult.getNoNextReason());
+    }
+
+    @Nonnull
+    private static RecordCursor<Integer> getFilteredFakeOutOfBandCursor(@Nonnull List<Integer> list, int limit, int threshold,
```

Comment:
The name here should probably reflect the fact that it returns on `OrElse` cursor

Reply:
Done.